Mathlib/Topology/MetricSpace/Ultra/Basic.lean

```diff
@@ -0,0 +1,235 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.MetricSpace.Pseudo.Lemmas
+
+/-!
+## Ultrametric spaces
+
+This file defines ultrametric spaces, implemented as a mixin on the `Dist`,
+so that it can apply on pseudometric spaces as well.
+
+## Main definitions
+
+* `IsUltrametricDist X`: Annotates `dist : X → X → ℝ` as respecting the ultrametric inequality
+  of `dist(x, z) ≤ max {dist(x,y), dist(y,z)}`
+
+## Implementation details
+
+The mixin could have been defined as a hypothesis to be carried around, instead of relying on
+typeclass synthesis. However, since we declare a (pseudo)metric space on a type using
+typeclass arguments, one can declare the ultrametricity at the same time.
+For example, one could say `[Norm K] [Fact (IsNonarchimedean (norm : K → ℝ))]`,
+
+The file imports a later file in the hierarchy of pseudometric spaces, since
+`Metric.isClosed_ball` and `Metric.isClosed_sphere` is proven in a later file
+using more conceptual results.
+
+TODO: Generalize to ultrametric uniformities
+
+## Tags
+
+ultrametric, nonarchimedean
+-/
+
+variable {X : Type*}
+
+/-- The `dist : X → X → ℝ` respects the ultrametric inequality
+of `dist(x, z) ≤ max (dist(x,y)) (dist(y,z))`. -/
+class IsUltrametricDist (X : Type*) [Dist X] : Prop where
+  dist_triangle_max : ∀ x y z : X, dist x z ≤ max (dist x y) (dist y z)
+
+open Metric
+
+variable [PseudoMetricSpace X] [IsUltrametricDist X] (x y z : X) (r s : ℝ)
+
+lemma dist_triangle_max : dist x z ≤ max (dist x y) (dist y z) :=
+  IsUltrametricDist.dist_triangle_max x y z
+
+namespace IsUltrametricDist
+
+lemma ball_eq_of_mem {x y : X} {r : ℝ} (h : y ∈ ball x r) : ball x r = ball y r := by
+  ext a
+  simp_rw [mem_ball] at h ⊢
+  constructor <;> intro h' <;>
+  exact (dist_triangle_max _ _ _).trans_lt (max_lt h' (dist_comm x _ ▸ h))
+
+lemma mem_ball_iff_mem_ball {x y: X} {r : ℝ} : y ∈ ball x r ↔ x ∈ ball y r := by
```

Comment:
Since we're in the namespace, I think we can get away with:
```suggestion
lemma mem_ball_iff {x y: X} {r : ℝ} : y ∈ ball x r ↔ x ∈ ball y r := by
```
but this is optional.